examples/rust/arkworks/src/main.rs

```diff
@@ -1,54 +1,333 @@
-    let ark_scalars = (0..size)
-        .map(|i| Fr::from(i as u64))
-        .collect();
-    return ark_scalars;
+fn incremental_ark_projective_points(size: usize) -> Vec<ArkProjective> {
+    (1..=size)
+        .map(|i| ArkAffine::generator().mul(&Fr::from(i as u64)))
+        .collect()
 }
 
-// Generic function to convert Arkworks field elements to Icicle format and return a mutable slice
+//============================================================================================//
+//========================= Convert single field element ark<->ICICLE ========================//
+//============================================================================================//
+fn from_ark<T, I>(ark: &T) -> I
+where
+    T: PrimeField,
+    I: FieldImpl,
+{
+    let mut ark_bytes = Vec::with_capacity(T::BigInt::NUM_LIMBS * 8 * T::extension_degree() as usize);
+    for base_elem in ark.to_base_prime_field_elements() {
+        ark_bytes.extend_from_slice(
+            &base_elem
+                .into_bigint()
+                .to_bytes_le(),
+        );
+    }
+    I::from_bytes_le(&ark_bytes)
+}
+
+fn to_ark<T, I>(icicle: &I) -> T
+where
+    T: PrimeField,
+    I: FieldImpl,
+{
+    T::from_random_bytes(&icicle.to_bytes_le()).unwrap()
+}
+
+//============================================================================================//
+//============================ Transmute or copy ark scalars =================================//
+//============================================================================================//
+
+// Generic function to transmute Arkworks field elements to Icicle format and return a mutable slice
 fn transmute_ark_to_icicle_scalars<T, I>(ark_scalars: &mut [T]) -> &mut [I]
 where
-    T: ark_ff::PrimeField,                // Arkworks field element must implement PrimeField
-    I: FieldImpl + MontgomeryConvertible, // Icicle scalar must implement a trait to convert from Arkworks type
+    T: PrimeField,
+    I: FieldImpl + MontgomeryConvertible,
 {
     // SAFETY: Reinterpreting Arkworks field elements as Icicle-specific scalars
     let icicle_scalars = unsafe { &mut *(&mut ark_scalars[..] as *mut _ as *mut [I]) };
 
-    // Create a HostSlice from the mutable slice
     let icicle_host_slice = HostSlice::from_mut_slice(&mut icicle_scalars[..]);
 
-    // Optional: Convert from Montgomery representation using the Icicle type's conversion method
+    // Convert from Montgomery representation using the Icicle type's conversion method
     I::from_mont(icicle_host_slice, &IcicleStream::default());
 
-    // Return the mutable slice of Icicle scalars
     icicle_scalars
 }
 
+fn ark_to_icicle_scalars_async<T, I>(ark_scalars: &[T], stream: &IcicleStream) -> DeviceVec<I>
+where
+    T: PrimeField,
+    I: FieldImpl + MontgomeryConvertible,
+{
+    // SAFETY: Reinterpreting Arkworks field elements as Icicle-specific scalars
+    let icicle_scalars = unsafe { &*(&ark_scalars[..] as *const _ as *const [I]) };
```

Comment:
Rust recommended to remove the double slicing here, since ark_scalars is already a slice. I think it allocates memory again when creating this extra slice. Minor change. 

fn ark_to_icicle_scalars_async<T, I>(ark_scalars: &[T], stream: &IcicleStream) -> DeviceVec<I>
where
    T: PrimeField,
    I: FieldImpl + MontgomeryConvertible,
{
    // SAFETY: Reinterpreting Arkworks field elements as Icicle-specific scalars
    let icicle_scalars = unsafe { &*(ark_scalars as *const _ as *const [I]) };

    // Create a HostSlice from the mutable slice
    let icicle_host_slice = HostSlice::from_slice(icicle_scalars);

    let mut icicle_scalars = DeviceVec::<I>::device_malloc_async(ark_scalars.len(), stream).unwrap();
    icicle_scalars
        .copy_from_host(icicle_host_slice)
        .unwrap();

    // Convert from Montgomery representation using the Icicle type's conversion method
    I::from_mont(&mut icicle_scalars, stream);
    icicle_scalars
}

Reply:
updated but it doesn't seem to be faster